docs/proposals/container-runtime-interface-v1.md

```diff
@@ -0,0 +1,188 @@
+  3. **Pod Spec is susceptible to inconsistent interpretations**. E.g., `rkt`
+     does not support container-level operations and assume immutable pods.
+  4. **Pod Spec is evolving rapidly**. New features are being added constantly.
+     Any pod-level change or addition requires changing of all container
+     runtime shims. E.g., init containers and volume containers.
+
+## Goals and Non-Goals
+
+The goals of defining the interface are to
+ - **improve extensibility**: Easier container runtime integration.
+ - **improve feature velocity**
+ - **improve code maintainability**
+
+The non-goals include
+ - proposing *how* to integrate with new runtimes, i.e., where the shim
+   resides. The discussion of adopting a client-server architecture is tracked
+   by [#13768](https://issues.k8s.io/13768), where benefits and shortcomings of
+   such an architecture is discussed.
+ - adding support to Windows containers. Windows container support is a
+   parallel effort and is tracked by [#22623](https://issues.k8s.io/22623). The
+   new interface will not be augmented to support Windows containers, but it will
+   be made extensible such that the support can be added in the future.
+ - re-defining Kubelet's internal interfaces. These interfaces, though, may affect
+   Kubelet's maintainability, is not relevant to runtime integration's.
+
+## Requirements
+
+ * Support existing integrated container runtimes
+ * Support hypervisor-based container runtimes
+
+## Container Runtime Interface
+
+The main idea of the proposal is to adopt an imperative container-level
+interface, which allows Kubelet to control the lifecycles of the containers
+inside a pod, across all container runtimes. A separate PodSandbox concept
+is defined to represent the environment in which a group of containers in a
+pod will be created and run. The container runtimes may interpret the sandbox
+concept differently based on how it operates internally. For runtimes relying
+on hypervisor, sandbox represents a virtual machine. For others, it can be
+Linux namespaces.
+
+In short, a PodSandbox should have the following features.
+
+ * **Isolation**: E.g., Linux namespaces or a full virtual machine, or even
+   support additional security features.
+ * **Compute resource Specifications**: A PodSandbox should implement pod-level
+   resource demands and restrictions.
+
+A container in a PodSandbox maps to a application in the Pod Spec. For Linux
+containers, they are expected to share at least network and IPC namespaces,
+with sharing more namespaces discussed in [#1615](https://issues.k8s.io/1615).
+
+
+Below is an example of the proposed interfaces.
+```go
+// PodSandboxManager contains basic operations for sandbox.
+type PodSandboxManager interface {
+	Create(config *PodSandboxConfig) (string, error)
+	Delete(id string) (string, error)
+	List(filter PodSandboxFilter) []PodSandboxListItem
+	Inspect(id string) PodSandbox
+    Metrics(id string) (PodSandboxMetrics, error)
```

Comment:
Keeping Metrics separate SGTM. It will let us plugin in the same provider
for multiple runtimes (cAdvisor)

On Tue, May 17, 2016 at 9:56 AM, Yu-Ju Hong <notifications@github.com>
wrote:

> In docs/proposals/container-runtime-interface-v1.md
> <https://github.com/yujuhong/kubernetes/pull/1#discussion_r63561043>:
>
> > +   resource demands and restrictions.
> > +
> > +A container in a PodSandbox maps to a application in the Pod Spec. For Linux
> > +containers, they are expected to share at least network and IPC namespaces,
> > +with sharing more namespaces discussed in [#1615](https://issues.k8s.io/1615).
> > +
> > +
> > +Below is an example of the proposed interfaces.
> > +```go
> > +// PodSandboxManager contains basic operations for sandbox.
> > +type PodSandboxManager interface {
> > +	Create(config *PodSandboxConfig) (string, error)
> > +	Delete(id string) (string, error)
> > +	List(filter PodSandboxFilter) []PodSandboxListItem
> > +	Inspect(id string) PodSandbox
> > +    Metrics(id string) (PodSandboxMetrics, error)
>
> The reason being that we may not always have pod-level cgroup. Container
> running on windows could be one example. That's also why I put CgroupParent
> in a Linux-only optional config.
>
> Initially I put metrics in a separate interface, which will allow better
> composability (i.e., runtime can implement this or not). @vishh
> <https://github.com/vishh> @dchen1107 <https://github.com/dchen1107>,
> what do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/yujuhong/kubernetes/pull/1/files/c0f4b02895d021d37b134e74a6ee9d5745fe982e#r63561043>
>